crates/autopilot/src/run_loop.rs

```diff
@@ -21,57 +25,105 @@ pub struct RunLoop {
 impl RunLoop {
     pub async fn run_forever(&self) -> ! {
         loop {
-            match self.single_run().await {
-                Ok(()) => tracing::debug!("single run ok"),
-                Err(err) => tracing::error!(?err, "single run err"),
-            }
+            self.single_run().await;
             tokio::time::sleep(Duration::from_secs(1)).await;
         }
     }
 
-    async fn single_run(&self) -> Result<()> {
+    async fn single_run(&self) {
         let auction = match self.solvable_orders_cache.current_auction() {
             Some(auction) => auction,
             None => {
                 tracing::debug!("no current auction");
-                return Ok(());
+                return;
+            }
+        };
+        let id = match self.database.replace_current_auction(&auction).await {
+            Ok(id) => id,
+            Err(err) => {
+                tracing::error!(?err, "failed to replace current auction");
+                return;
             }
         };
-        let id = self
-            .database
-            .replace_current_auction(&auction)
-            .await
-            .context("replace_current_auction")?;
+        self.single_run_(id, &auction)
+            .instrument(tracing::info_span!("auction", id))
+            .await;
+    }
 
-        let _responses = self.solve(&auction, id).await;
+    async fn single_run_(&self, id: AuctionId, auction: &Auction) {
+        tracing::info!("solving");
+        let mut solutions = self.solve(auction, id).await;
 
-        // TODO:
-        // - find winner
-        // - send execute to winner
-        // - store per auction info
+        // Shuffle so that sorting randomly splits ties.
+        solutions.shuffle(&mut rand::thread_rng());
+        solutions.sort_unstable_by(|left, right| left.1.objective.total_cmp(&right.1.objective));
 
-        Ok(())
+        // TODO: Keep going with other solutions until some deadline.
+        if let Some((index, solution)) = solutions.pop() {
+            tracing::info!("executing with solver {}", index);
+            match self
+                .execute(auction, id, &self.drivers[index], &solution)
+                .await
+            {
+                Ok(()) => tracing::info!("settled"),
+                Err(err) => {
+                    tracing::error!(?err, "solver {} failed to execute", err);
```

Comment:
I imagine you meant something more like this? Otherwise error is duplicated.
```suggestion
                    tracing::error!(?err, "solver {index} failed to execute");
```

Reply:
Good catch